js/data/symbol_bucket.js

```diff
@@ -36,10 +36,16 @@ function SymbolBucket(options) {
     var zoomHistory = { lastIntegerZoom: Infinity, lastIntegerZoomTime: 0, lastZoom: 0 };
 
     this.adjustedTextMaxSize = this.layer.getLayoutValue('text-size', 18, zoomHistory);
-    this.adjustedTextSize = this.layer.getLayoutValue('text-size', this.zoom + 1, zoomHistory);
+    this.adjustedTextSizes = [
+        this.layer.getLayoutValue('text-size', this.zoom, zoomHistory),
+        this.layer.getLayoutValue('text-size', this.zoom + 1, zoomHistory)
+    ];
 
     this.adjustedIconMaxSize = this.layer.getLayoutValue('icon-size', 18, zoomHistory);
-    this.adjustedIconSize = this.layer.getLayoutValue('icon-size', this.zoom + 1, zoomHistory);
+    this.adjustedIconSizes = [
+        this.layer.getLayoutValue('icon-size', this.zoom, zoomHistory),
+        this.layer.getLayoutValue('icon-size', this.zoom + 1, zoomHistory)
+    ];
```

Comment:
It is a simple algorithm. It really ought to go in the `mapbox-gl-function` repo but I'd look the other way. 

```
Math.max(at(minZoom), at(maxZoom), at(all stops between minzoom and maxzoom)
```

I have too much on my plate to tackle this in the next couple days. I think you should do it. Its 3 - 4 lines of code that'll 

 - be the fully correct solution to the problem
 - eliminate all known edge cases
 - expose an API that might be useful in the future
 - hedge against ever having to look at this particular code again